argon2.go

```diff
@@ -0,0 +1,161 @@
+		return nil, errors.New("MemoryKiB too large")
+	case o.Parallel > math.MaxUint8:
+		return nil, errors.New("Parallel too large")
+	case o.ForceIterations < 0:
+		return nil, errors.New("ForceIterations can't be negative")
+	case int64(keyLen) > math.MaxUint32:
+		return nil, errors.New("keyLen too large")
+	case o.ForceIterations > 0:
+		threads := runtimeNumCPU()
+		if threads > 4 {
+			threads = 4
+		}
+		params := &CostParams{
+			Time:      uint32(o.ForceIterations),
+			MemoryKiB: 1 * 1024 * 1024,
+			Threads:   uint8(threads)}
+
+		if o.MemoryKiB != 0 {
+			params.MemoryKiB = uint32(o.MemoryKiB)
+		}
+		if o.Parallel != 0 {
+			params.Threads = uint8(o.Parallel)
+			if o.Parallel > 4 {
+				params.Threads = 4
+			}
+		}
+
+		return params, nil
+	default:
+		benchmarkParams := &argon2.BenchmarkParams{
+			MaxMemoryCostKiB: 1 * 1024 * 1024,
+			TargetDuration:   2 * time.Second}
+
+		if o.MemoryKiB != 0 {
+			benchmarkParams.MaxMemoryCostKiB = uint32(o.MemoryKiB)
+		}
+		if o.TargetDuration != 0 {
+			benchmarkParams.TargetDuration = o.TargetDuration
+		}
+		if o.Parallel != 0 {
+			benchmarkParams.Threads = uint8(o.Parallel)
+			if o.Parallel > 4 {
+				benchmarkParams.Threads = 4
+			}
+		}
+
+		params, err := argon2.Benchmark(benchmarkParams, func(params *CostParams) (time.Duration, error) {
+			return kdf.Time(params, uint32(keyLen))
+		})
+		if err != nil {
+			return nil, xerrors.Errorf("cannot benchmark KDF: %w", err)
+		}
+
+		return params, nil
+	}
+}
+
+// CostParams defines the cost parameters for key derivation using Argon2.
+type CostParams = argon2.CostParams
+
+// KDF is an interface to abstract use of the Argon2 KDF to make it possible
+// to delegate execution to a short-lived utility process where required.
+type KDF interface {
+	// Derive derives a key of the specified length in bytes, from the supplied
+	// passphrase and salt and using the supplied cost parameters.
+	Derive(passphrase string, salt []byte, params *CostParams, keyLen uint32) ([]byte, error)
+
+	// Time measures the amount of time the KDF takes to execute with the
+	// specified cost parameters and key length in bytes.
+	Time(params *CostParams, keyLen uint32) (time.Duration, error)
+}
+
+type argon2iKDFImpl struct{}
+
+func (_ argon2iKDFImpl) Derive(passphrase string, salt []byte, params *CostParams, keyLen uint32) ([]byte, error) {
+	return argon2.Key(passphrase, salt, params, keyLen), nil
+}
+
+func (_ argon2iKDFImpl) Time(params *CostParams, keyLen uint32) (time.Duration, error) {
+	return argon2.KeyDuration(params, keyLen), nil
+}
+
+var argon2iKDF = argon2iKDFImpl{}
+
+// Argon2iKDF returns the in-process Argon2i implementation of KDF. This
+// shouldn't be used in long-lived system processes - these processes should
+// instead provide their own KDF implementation which delegates to a short-lived
+// utility process which will use the in-process implementation.
+func Argon2iKDF() KDF {
+	return argon2iKDF
+}
```

Comment:
this are trivial but nevertheless not tested

Reply:
I've added test cases for this now